Add a bitwise NOT operator for integer and Boolean types (CPU).

xuhdev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #22326 Delegate Python ~ (invert operator) to Tensor.bitwise_not().
* #22320 Add a bitwise NOT operator for integer and Boolean types (CUDA).
* **#22283 Add a bitwise NOT operator for integer and Boolean types (CPU).**

Related to #20024, #22246, #22262

Differential Revision: [D16183576](https://our.internmc.facebook.com/intern/diff/D16183576)

xuhdev committed:
Add a bitwise NOT operator for integer and Boolean types (CPU).

Related to #20024, #22246, #22262

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CPU)."

Add a bitwise NOT operator for integer and Boolean types (CPU).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22283 gh/xuhdev/6/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CPU)."

Add a bitwise NOT operator for integer and Boolean types (CPU).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22283 gh/xuhdev/6/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CPU)."

Add a bitwise NOT operator for integer and Boolean types (CPU).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22283 gh/xuhdev/6/head

xuhdev commented:
In subsequent PRs, I will implement a GPU version and replace the `~` operator implementation with this function. But this PR by itself can also be seen as an independent contribution.

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CPU)."

Add a bitwise NOT operator for integer and Boolean types (CPU).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22283 gh/xuhdev/6/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CPU)."

Add a bitwise NOT operator for integer and Boolean types (CPU).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22283 gh/xuhdev/6/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CPU)."

Add a bitwise NOT operator for integer and Boolean types (CPU).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22283 gh/xuhdev/6/head

izdeby commented on aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:
```diff
@@ -51,6 +51,28 @@ static void abs_kernel(TensorIterator& iter) {
   });
 }
 
+// Boolean type does not work with ~ (bitwise NOT) in C++. bitwise_not_int_bool wraps this operation for both Boolean
```
what is bitwise_not_int_bool?

xuhdev replied:
Updated

izdeby commented on test/test_torch.py:
```diff
@@ -1748,6 +1748,25 @@ def _test_neg(self, cast):
 
+    def test_bitwise_not(self):
+        res = 0xffff - torch.arange(127, dtype=torch.int8)
+        for t in (torch.BoolTensor, torch.LongTensor, torch.IntTensor, torch.ShortTensor, torch.CharTensor):
```
should here be also torch.ByteTensor?

xuhdev replied:
Yes. Added.

izdeby commented on test/test_torch.py:
```diff
@@ -1748,6 +1748,25 @@ def _test_neg(self, cast):
+        for t in (torch.BoolTensor, torch.LongTensor, torch.IntTensor, torch.ShortTensor, torch.CharTensor):
+            if t == torch.BoolTensor:
+                a = torch.BoolTensor([True, False])
+                this_res = torch.BoolTensor([False, True])
```
nit: 'expectedRes' or similar instead of this_res?

xuhdev replied:
I changed to `expected_this`. I meant "result in this iteration".

izdeby commented on torch/_torch_docs.py:
```diff
@@ -704,6 +704,22 @@ def merge_dicts(*dicts):
     tensor([0.0000, 0.0000, 0.0000, 1.0000, 1.0000, 0.0000, 0.5000])
 """)
 
+add_docstr(torch.bitwise_not,
```
doc for inplace version of bitwise_not?

xuhdev replied:
Added.

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CPU)."

Add a bitwise NOT operator for integer and Boolean types (CPU).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22283 gh/xuhdev/6/head

izdeby commented:
are you going to add support for CUDA?

xuhdev commented:
@izdeby Yes

colesbury commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -339,6 +339,18 @@
+  variants: function, method
+
+- func: bitwise_not_(Tensor(a!) self) -> Tensor(a!)
+  variants: function, method
```
The in-place version should be method-only

xuhdev replied:
Hmm... There are several other in-place versions with `function` variant. Should they also be changed?

xuhdev replied:
For example, `abs_`, `ceil_`

colesbury replied:
Yes, they should probably be changed (but not as part of this PR)

cc @gchanan

xuhdev replied:
Sure -- Just to confirm I need to make these changes. Updated.

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CPU)."

Add a bitwise NOT operator for integer and Boolean types (CPU).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22283 gh/xuhdev/6/head

colesbury reviewed (dismissed):
lgtm too

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CPU)."

Add a bitwise NOT operator for integer and Boolean types (CPU).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22283 gh/xuhdev/6/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CPU)."

Add a bitwise NOT operator for integer and Boolean types (CPU).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22283 gh/xuhdev/6/head

xuhdev commented on test/test_torch.py:
```diff
@@ -1748,6 +1748,40 @@ def _test_neg(self, cast):
+            self.assertEqual(expected_res, a)
+
+        # test exceptions
+        for t in(torch.HalfTensor, torch.FloatTensor, torch.DoubleTensor):
```
@izdeby I added more tests for exceptions here

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CPU)."

Add a bitwise NOT operator for integer and Boolean types (CPU).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22283 gh/xuhdev/6/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CPU)."

Add a bitwise NOT operator for integer and Boolean types (CPU).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22283 gh/xuhdev/6/head

xuhdev commented:
@pytorchbot retest this please

xuhdev commented:
@pytorchbot retest this please

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CPU)."

Add a bitwise NOT operator for integer and Boolean types (CPU).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22283 gh/xuhdev/6/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CPU)."

Add a bitwise NOT operator for integer and Boolean types (CPU).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22283 gh/xuhdev/6/head

xuhdev commented:
@colesbury @izdeby I changed the code to avoid the template `bitwise_not` function (and thus the code becomes simpler)

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CPU)."

Add a bitwise NOT operator for integer and Boolean types (CPU).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22283 gh/xuhdev/6/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CPU)."

Add a bitwise NOT operator for integer and Boolean types (CPU).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22283 gh/xuhdev/6/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CPU)."

Add a bitwise NOT operator for integer and Boolean types (CPU).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22283 gh/xuhdev/6/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CPU)."

Add a bitwise NOT operator for integer and Boolean types (CPU).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22283 gh/xuhdev/6/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CPU)."

Add a bitwise NOT operator for integer and Boolean types (CPU).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22283 gh/xuhdev/6/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CPU)."

Add a bitwise NOT operator for integer and Boolean types (CPU).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22283 gh/xuhdev/6/head

zou3519 commented:
@izdeby @colesbury could you take another look at this please?

izdeby commented on test/test_torch.py:
```diff
@@ -1748,6 +1748,40 @@ def _test_neg(self, cast):
 
+    def test_bitwise_not(self):
+        res = 0xffff - torch.arange(127, dtype=torch.int8)
+        for t in (torch.BoolTensor,
```
nit: can you, please, change variable names to something more readable? for example, 'dtype' instead of 't' and so on.

xuhdev replied:
Yes. I made the corresponding changes in #22320 and #22326 as those are the places where dtype is used.

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CPU)."

Add a bitwise NOT operator for integer and Boolean types (CPU).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22283 gh/xuhdev/6/head
